backend/src/org/commcare/suite/model/OfflineUserRestore.java

```diff
@@ -0,0 +1,135 @@
+    }
+
+    public static OfflineUserRestore buildInMemoryUserRestore(InputStream restoreStream) throws IOException {
+        OfflineUserRestore offlineUserRestore = new OfflineUserRestore();
+        offlineUserRestore.restore = new String(StreamsUtil.inputStreamToByteArray(restoreStream));
+        return offlineUserRestore;
+    }
+
+    public InputStream getRestoreStream() {
+        if (reference == null) {
+            return getInMemoryStream();
+        } else {
+            return getStreamFromReference();
+        }
+    }
+
+    private InputStream getStreamFromReference() {
+        try {
+            Reference local = ReferenceManager._().DeriveReference(reference);
+            return local.getStream();
+        } catch (IOException | InvalidReferenceException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private InputStream getInMemoryStream() {
+        try {
+            return new ByteArrayInputStream(restore.getBytes("UTF-8"));
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public String getReference() {
+        return reference;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public String getPassword() {
+        return password;
+    }
+
+    @Override
+    public void readExternal(DataInputStream in, PrototypeFactory pf)
+            throws IOException, DeserializationException {
+        this.recordId = ExtUtil.readInt(in);
+        this.reference = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+        this.restore = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+    }
+
+    @Override
+    public void writeExternal(DataOutputStream out) throws IOException {
+        ExtUtil.writeNumeric(out, recordId);
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(reference));
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(restore));
+    }
+
+    @Override
+    public void setID(int ID) {
+        recordId = ID;
+    }
+
+    @Override
+    public int getID() {
+        return recordId;
+    }
+
+    private static String parseUsernameFromRestoreFile(String localRestoreReference) {
+        try {
+            InputStream is = ReferenceManager._().DeriveReference(localRestoreReference).getStream();
+            KXmlParser parser = ElementParser.instantiateParser(is);
+            parser.next();
+            int eventType = parser.getEventType();
+            do {
+                if (eventType == KXmlParser.START_TAG) {
+                    if (parser.getName().toLowerCase().equals("username")) {
+                        parser.next();
+                        return parser.getText();
+                    }
+                }
+                eventType = parser.next();
+            } while (eventType != KXmlParser.END_DOCUMENT);
+        } catch (IOException | XmlPullParserException | InvalidReferenceException e) {
+            return null;
```

Comment:
Mostly fixed this up, will finalize once we decide to what extent we're parsing the restore file at install time